azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusUpdatingListener.java

```diff
@@ -0,0 +1,317 @@
+  }
+
+  /**
+   * Apply the boolean function {@code expectedStatusMatcher} to execution-id and if it returns true
+   * then finalize the status of the flow in db.
+   * This can be used for finalizing the flow based on whether the current Flow Status has a
+   * specific value or is present within a given set of values.
+   *
+   * <p> Note:
+   * Unfortunately many (if not all) of the Flow status updates in Azkaban don't fully enforce
+   * the Flow lifecycle state-machine and {@code ExecutionControllerUtils.finalizeFlow} used
+   * within this method is no exception. It will simply finalize the flow (as failed) even if the
+   * flow status is in a finalized state in the Db. <br>
+   * This indirectly impacts this listener implementation as occasionally more than one thread
+   * could try to update the state of the same flow in db. Although it's not any worse than how
+   * the rest of Azkaban already behaves, we should fix the behavior at least during the
+   * finalization of flows. One way of achieving this is to add a utility method to atomically
+   * test-and-finalize a flow from a non-final to a failed state.
+   *
+   * @implNote Flow status check and update is not atomic, details above.
+   * @param event
+   * @param expectedStatusMatcher
+   * @return
+   */
+  private azkaban.executor.Status compareAndFinalizeFlowStatus(AzPodStatusMetadata event,
+      Function<Status, Boolean> expectedStatusMatcher) {
+    requireNonNull(expectedStatusMatcher, "flow matcher must not be null");
+
+    int executionId = Integer.parseInt(event.getFlowPodMetadata().get().getExecutionId());
+    ExecutableFlow executableFlow = null;
+    try {
+      executableFlow = executorLoader.fetchExecutableFlow(executionId);
+    } catch (ExecutorManagerException e) {
+      String message = format("Exception while fetching executable flow for pod %s",
+          event.getPodName());
+      logger.error(message, e);
+      throw new AzkabanWatchException(message, e);
+    }
+    checkState(executableFlow != null, "executable flow must not be null");
+    azkaban.executor.Status originalStatus = executableFlow.getStatus();
+
+    if (!expectedStatusMatcher.apply(originalStatus)) {
+      logger.info(format(
+          "Flow for pod %s does not have the expected status in database and will be finalized.",
+          event.getPodName()));
+      final String reason = "Flow Pod execution was completed.";
+      ExecutionControllerUtils.finalizeFlow(executorLoader, alerterHolder, executableFlow, reason,
+          null);
+    }
+    return originalStatus;
+  }
+
+  /**
+   * Delete the the flow pod and any other related objects (such as services).
+   *
+   * @param event pod watch event
+   */
+  private void deleteFlowContainer(AzPodStatusMetadata event) {
+    logger.info("Deleting Flow Pod: " + event.getPodName());
+    try {
+      containerizedImpl.deleteContainer(
```

Comment:
Just wondering if deleteContainer should take care of validating flow status and set appropriately if it is not end state.

Reply:
Wanted keep this function both small and specific to the objective of deleting a pod. Any additional checks for flows status etc can be composed in a higher level method like onPodCompleted